include/hip/hcc_detail/program_state.hpp

```diff
@@ -399,29 +397,46 @@ const std::vector<hsa_executable_t>& executables(hsa_agent_t agent) {
 
     std::call_once(r[agent].first, [](hsa_agent_t aa) {
         hsa_agent_iterate_isas(aa, [](hsa_isa_t x, void* pa) {
-            const auto it = code_object_blobs().find(x);
-
-            if (it == code_object_blobs().cend()) return HSA_STATUS_SUCCESS;
-
-            hsa_agent_t a = *static_cast<hsa_agent_t*>(pa);
-
-            for (auto&& blob : it->second) {
-                hsa_executable_t tmp = {};
+            for (const auto code_object_it : code_object_blobs()) {
+                const auto elf = code_object_it.first;
+                const auto it = code_object_it.second.find(x);
 
-                hsa_executable_create_alt(
-                    HSA_PROFILE_FULL,
-                    HSA_DEFAULT_FLOAT_ROUNDING_MODE_DEFAULT,
-                    nullptr,
-                    &tmp);
+                if (it == code_object_it.second.cend()) continue;
 
-                // TODO: this is massively inefficient and only meant for
-                // illustration.
-                std::string blob_to_str{blob.cbegin(), blob.cend()};
-                tmp = load_executable(blob_to_str, tmp, a);
+                hsa_agent_t a = *static_cast<hsa_agent_t*>(pa);
 
-                if (tmp.handle) r[a].second.push_back(tmp);
+                std::vector<hsa_executable_t> current_exes;
+                // check the cache for already loaded executables
+                hip_impl::executables_cache(elf, x, a, current_exes, false/*read, not write*/);
+                if (!current_exes.empty()) {
+                    // found already loaded, append and continue with next elf
+                    r[a].second.insert(r[a].second.end(), current_exes.begin(), current_exes.end());
+                    continue;
+                }
+                // executables do not yet exist for this elf+isa+agent, create and cache them
+                for (auto&& blob : it->second) {
+                    hsa_executable_t tmp = {};
+
+                    hsa_executable_create_alt(
+                        HSA_PROFILE_FULL,
+                        HSA_DEFAULT_FLOAT_ROUNDING_MODE_DEFAULT,
+                        nullptr,
+                        &tmp);
+
+                    // TODO: this is massively inefficient and only meant for
+                    // illustration.
+                    std::string blob_to_str{blob.cbegin(), blob.cend()};
+                    tmp = load_executable(blob_to_str, tmp, a);
+
+                    if (tmp.handle) {
+                        current_exes.push_back(tmp);
+                    }
+                }
+                // cache the newly loaded executables
+                hip_impl::executables_cache(elf, x, a, current_exes, true/*write, not read*/);
```

Comment:
I think the retrieval of the cached map, the load of a new executable and the cache update should be all in a critical section protected by a global mutex (since the cache is global).  I don't think the call_once flag is sufficient here to protect against another thread running the same code but for a different module.